src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java

```diff
@@ -76,93 +92,166 @@ private void deleteWebhooks(List<BitbucketWebhook> webhooks) {
+     * @return the correct registered/updated bitbucket webhook
+     */
+    @Nullable
     private BitbucketWebhook process(WebhookRegisterRequest request,
-                                     BitbucketWebhookEvent event) {
+                                     Collection<BitbucketWebhookEvent> events) {
         String callback = constructCallbackUrl(request);
         List<BitbucketWebhook> ownedHooks =
-                webhookClient.getWebhooks(REPO_REF_CHANGE.getEventId(), MIRROR_SYNCHRONIZED_EVENT.getEventId())
+                webhookClient.getWebhooks(getEventIdAsStrings(MIRROR_SYNCHRONIZED, PULL_REQUEST_DECLINED,
+                        PULL_REQUEST_DELETED, PULL_REQUEST_FROM_REF_UPDATED, PULL_REQUEST_MERGED,
+                        PULL_REQUEST_OPENED, REPO_REF_CHANGE))
                         .filter(hook -> hook.getName().equals(request.getName()) || hook.getUrl().equals(callback))
                         .collect(toList());
+
+        //Creating lists of different kinds of webhooks
         List<BitbucketWebhook> webhookWithMirrorSync = ownedHooks.stream()
-                .filter(hook -> hook.getEvents().contains(MIRROR_SYNCHRONIZED_EVENT.getEventId()))
+                .filter(hook -> hook.getEvents().contains(MIRROR_SYNCHRONIZED.getEventId()))
                 .collect(toList());
-        List<BitbucketWebhook> webhookWithRepoRefChange = ownedHooks
+        List<BitbucketWebhook> webhookWithRepoRefChangeOnly = ownedHooks
                 .stream()
-                .filter(hook -> hook.getEvents().contains(REPO_REF_CHANGE.getEventId()))
+                .filter(hook -> hook.getEvents().equals(Collections.singleton(REPO_REF_CHANGE.getEventId())))
                 .collect(toList());
 
-        if (ownedHooks.size() == 0 ||
-            (webhookWithMirrorSync.size() == 0 && event == MIRROR_SYNCHRONIZED_EVENT) ||
-            (webhookWithRepoRefChange.size() == 0 && event == REPO_REF_CHANGE)) {
-            BitbucketWebhookRequest webhook = createRequest(request, event);
+        List<BitbucketWebhook> webhookWithPROnly = ownedHooks
+                .stream()
+                .filter(hook -> hook.getEvents().equals(ALL_PULL_REQUEST_EVENT_IDS))
+                .collect(toList());
+
+        List<BitbucketWebhook> webhookWithRepoRefChangeOrPR = ownedHooks
+                .stream()
+                .filter(hook -> hook.getEvents().stream().anyMatch(refAndPREventIds::contains))
+                .collect(toList());
+
+        //Determining whether to create a new webhook
+        //We will handle mirror syncs separate from all other events.
+        // If no webhook with wanted events exist, we register a new one.
+        if (ownedHooks.isEmpty() ||
+            (webhookWithMirrorSync.isEmpty() && events.contains(MIRROR_SYNCHRONIZED)) ||
+            (webhookWithRepoRefChangeOrPR.isEmpty() && (events.contains(PULL_REQUEST_OPENED) || events.contains(REPO_REF_CHANGE)))) {
```

Comment:
Same as below, this isn't checking source branch updated (though I don't know if that would actually lead to an issue)